hudi-common/src/main/java/org/apache/hudi/common/table/read/BufferedRecordSerializer.java

```diff
@@ -31,52 +31,81 @@
 
 /**
  * An implementation of {@link CustomSerializer} for {@link BufferedRecord}.
- *
  */
 public class BufferedRecordSerializer<T> implements CustomSerializer<BufferedRecord<T>> {
-  public static final int KRYO_SERIALIZER_INITIAL_BUFFER_SIZE = 1048576;
-  private final Kryo kryo;
-  // Caching ByteArrayOutputStream to avoid recreating it for every operation
-  private final ByteArrayOutputStream baos;
+  // Caching kryo serializer to avoid creating kryo instance for every serde operation
+  private static final ThreadLocal<InternalSerializerInstance> SERIALIZER_REF =
```

Comment:
curious to understand the overhead here. 
we will be instantiating just once per file group right? (even before this patch). I see we use this in ExternalSpillableMap for the merged log records and so its one instance per file group. 
so, how much we might get a benefit in here. 
and where do we access these across threads. Isn't Threadlocal.withInitial() mainly helps w/ thread safety across threads and each thread gets its own local copy. 

can you shed some light please.

Reply:
This is using ThreadLocal so that each thread will have its own copy. You don't need to make new copies per usage on that thread as a result.

In ExternalSpillableMap we used `SerializationUtils` in the past which uses the same approach. The approach is aligned with the [kryo docs](https://github.com/EsotericSoftware/kryo?tab=readme-ov-file#pooling)